an-rope/src/internals.rs

```diff
@@ -45,101 +44,74 @@ fn fibonacci(n: usize) -> usize {
 impl BranchNode {
 
     #[inline]
-    fn new(left: Box<Node>, right: Box<Node>) -> Self {
+    fn new(left: Node, right: Node) -> Self {
         BranchNode { len: left.len() + right.len()
                    , weight: left.subtree_weight()
-                   , left: Some(left)
-                   , right: Some(right)
+                   , left: box left
+                   , right: box right
                    }
     }
 
-    fn split(&mut self, index: usize) -> &mut Node {
-        if index < self.weight {
+    #[inline]
+    fn take_left(&mut self) -> Node {
+        mem::replace(self.left.as_mut(), Node::empty())
+    }
+
+    #[inline]
+    fn take_right(&mut self) -> Node {
+        mem::replace(self.right.as_mut(), Node::empty())
+    }
+
+    fn split(self, index: usize) -> (Node, Node) {
+        let weight = (&self).weight;
+        if index < weight {
             // split the left node
-            *self = if let BranchNode {
-                left: Some(box ref mut left)
-              , ref mut right, ..
-            } = *self {
-                let mut l_leaf = mem::replace(left, Node::empty());
-                l_leaf.split(index);
-                let r = right.take().unwrap_or(box Node::empty());
-                // replacing *self with a new BranchNode will update the
-                // node's weight automagically
-                BranchNode::new(box l_leaf, r)
+            let (left, left_right) = self.left.split(index);
```

Comment:
This naming is sorta confusing and the comment is not very helpful, obviously `foo.split()` splits foo :)

I'd prefer it if the comments in the branches explained what the intent of the operation was, i.e. "split the left node in two new nodes: `foo` which contains blah and `bar` which contains bleh. Then make a new branch containing `foo` and `bar` as children and ..."